### PR TITLE
Added K Nearest Neighbours Algorithm in C++

### PR DESCRIPTION
**Pull request by Aman-Codes**
**Fixes issue:** #4634 

**Changes:**

Added K Nearest Neighbours Algorithm in C++

### CODE REVIEW

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
+// Part of Cosmos by OpenGenus
+#include <bits/stdc++.h>
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: #1915. Don't use using namespace std either.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+using namespace std;
+
+struct point
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Capital P.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    float w, x, y, z; // w, x, y and z are the values of the dataset
+    float distance; // distance will store the distance of dataset point from the unknown test point
+    int op; // op will store the class of the point
+};
+
+int main() 
+{
+    int i, j, n, k;
+    struct point p;
+    struct point temp;
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]:
```suggestion
    point p;
    point temp;
```

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    cout << "Number of data points: ";
+    cin >> n;
+    struct point arr[n];
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Use `std::vector` instead.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    cout << "\nEnter the dataset values: \n";
+    cout << "w\tx\ty\tz\top\n";
+
+    // Taking dataset
+    for (i = 0; i < n; i++)
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Declare i inside the loop itself.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        cout << fixed << setprecision(2) << arr[i].w << '\t' << arr[i].x << '\t' << arr[i].y << '\t' << arr[i].z << '\t' << arr[i].op << '\n';
+    }
+
+    cout << "\nEnter the feature values of the unknown point: \nw\tx\ty\tz\n";
+    cin >> p.w >> p.x >> p.y >> p.z;
+
+    // Measuring the Euclidean distance
+    for (i = 0; i < n; i++) 
+    {
+        arr[i].distance = sqrt(((arr[i].w - p.w) * (arr[i].w - p.w)) + ((arr[i].x - p.x) * (arr[i].x - p.x)) + ((arr[i].y - p.y) * (arr[i].y - p.y)) + ((arr[i].z - p.z) * (arr[i].z - p.z)));
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Extract this to a function, such as `euclideanDistance`.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    // Sorting the training data with respect to distance
+    for (i = 1; i < n; i++)
+    {
+        for (j = 0; j < n - i; j++) 
+        {
+            if (arr[j].distance > arr[j + 1].distance)
+            {
+                temp = arr[j];
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Declare temp inside this conditional.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    for (int i = 0; i < 1000; i++) 
+    {
+        if(freq[i] > maxfreq) 
+        {
+            maxfreq = freq[i];
+            index = i;
+        }
+    }
+
+    cout << "The class of unknown point is " << index;
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]:
```suggestion
    cout << "The class of the unknown point is " << index;
```

**Comment by arnavb** [comment]
Finally, reformat your code using clang-format.

**Comment by Aman-Codes** [comment]
@arnavb Sir did the asked changes.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
+// Part of Cosmos by OpenGenus
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Filename should be all lowercase.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+#include <iostream>
+#include <iomanip>
+#include <math.h>
+#include<vector>
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Reformat your file using clang-format.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
+// Part of Cosmos by OpenGenus
+#include <iostream>
+#include <iomanip>
+#include <math.h>
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Use cmath instead.

File: code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        vec[i].w = wCoordinate;
+        vec[i].x = xCoordinate;
+        vec[i].y = yCoordinate;
+        vec[i].z = zCoordinate;
+        vec[i].op = opCoordinate;
+    }
+
+    std::cout << "\nw\tx\ty\tz\top\n";
+
+    for (int i = 0; i < n; i++)
```
Comment (arnavb) [inline code/artificial_intelligence/src/k_nearest_neighbours/K_Nearest_Neighbours.cpp]: Use pre-increment instead of post in all loops.

**Comment by Aman-Codes** [comment]
@arnavb  Sir can u please guide me how to install clang-format for windows and use it to format c++ code. I am facing some issues in it.

**Comment by arnavb** [comment]
@Aman-Codes You don't have to install it locally. There is clang format online [here](https://zed0.co.uk/clang-format-configurator/). Just set the indentation width to 4 spaces.

**Comment by Aman-Codes** [comment]
@arnavb Sir did the asked changes and formatted the code by clang format. Do let me know if any changes are required.

**Pull request merged**